pkg/component/controller/cplb/cplb_label_linux.go

```diff
@@ -0,0 +1,128 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cplb
+
+import (
+	"fmt"
+	"net"
+	"unsafe"
+
+	"github.com/vishvananda/netlink/nl"
+	"golang.org/x/sys/unix"
+)
+
+// maybeSetAddrLabels sets the labels for the vips to vrrp.AddressLabel
+// so that the real IP is preferred.
+func (k *Keepalived) maybeSetAddrLabels() error {
+	for _, vrrp := range k.Config.VRRPInstances {
+		for _, vip := range vrrp.VirtualIPs {
+			// Only set labels for IPv6 addresses
+			ipAddr, _, err := net.ParseCIDR(vip)
+			if err != nil {
+				return fmt.Errorf("failed to parse CIDR %s: %w", vip, err)
+			}
+
+			// Only set labels for IPv6 addresses
+			if ipAddr.To4() != nil {
+				continue
+			}
+
+			// Set address label for IPv6 VIP
+			if err := setAddressLabel(ipAddr, 128, vrrp.AddressLabel); err != nil {
+				return fmt.Errorf("failed to set address label for %s: %w", ipAddr, err)
+			}
+		}
+	}
+	return nil
+}
+
+func setAddressLabel(ip net.IP, bits int, label uint32) error {
+	// Use iproute2 as reference for setting the address label.
+	// https://kernel.googlesource.com/pub/scm/network/iproute2/iproute2/+/refs/tags/v6.15.0/ip/ipaddrlabel.c#176
+	if ip.To4() != nil {
+		return fmt.Errorf("cannot set address label for IPv4 address %s", ip)
+	}
+	req := nl.NewNetlinkRequest(unix.RTM_NEWADDRLABEL,
+		unix.NLM_F_REQUEST|unix.NLM_F_ACK|unix.NLM_F_REPLACE|unix.NLM_F_CREATE)
+
+	msg := &IfAddrLabelmsg{
+		Family:    uint8(unix.AF_INET6),
+		PrefixLen: uint8(bits),
+	}
+
+	req.AddData(msg)
+
+	// RFC 3484 and 6724 don't specify label size, but iproute2 uses 32 bit labels:
+	// https://kernel.googlesource.com/pub/scm/network/iproute2/iproute2/+/refs/tags/v6.15.0/ip/ipaddrlabel.c#176
+	labelBytes := make([]byte, 4)
+	nl.NativeEndian().PutUint32(labelBytes, label)
+
+	labelData := nl.NewRtAttr(IFAL_LABEL, labelBytes)
+	req.AddData(labelData)
+
+	// Set IFAL_ADDRESS
+	addressData := nl.NewRtAttr(IFAL_ADDRESS, []byte(ip.To16()))
+	req.AddData(addressData)
+
+	_, err := req.Execute(unix.NETLINK_ROUTE, 0)
+	return err
+}
+
+// Everything below this line should be defined in "golang.org/x/sys/unix"
+// and "github.com/vishvananda/netlink/nl" packages.
+// Ideally we'll get it merged into those packages in the future, but for now
+// define it here.
+
+type IfAddrLabelmsg struct {
+	Family    uint8
+	_         uint8
+	PrefixLen uint8
+	Flags     uint8
+	Index     uint32
+	Seq       uint32
+}
+
+const (
+	IFAL_ADDRESS = 1
+	IFAL_LABEL   = 2
+)
+
+const SizeOfIfAddrLabelMsg = 12 // Size of IfAddrLabelMsg in bytes
```

Comment:
```suggestion
const sizeOfIfAddrLabelMsg = int(unsafe.Sizeof(IfAddrLabelmsg{}))
```

Reply:
We can do that, I did it this way because the code of [sys/unix](https://github.com/golang/sys/blob/master/unix/ztypes_linux.go#L688C1-L688C26) just uses a constant value. But if you want I can change that.

Anyway I'm working on adding it to upstream go once I figure out the sys/unix code generation.

Reply:
I really think we should use the unsafe.Sizeof version. Otherwise, we should add a unit test that asserts that asserts `sizeOfIfAddrLabelMsg = int(unsafe.Sizeof(IfAddrLabelmsg{}))`. Even if this code seems to be intermediary until upstream changes land, you'll never know how long it will stay, and whenever the struct size changes for whichever reason, you'd be on your own to remember that this constant needs to be changed.

I don't think the constant needs to be public, too.